File: basic_dsa_two_hundred_fifty.java
/**
 * Basic DSA Problems - Heaters
 * Find minimum radius so all houses can be warmed by heaters
 */
public class basic_dsa_two_hundred_fifty {

    // Problem: houses[] and heaters[] positions. Each heater warms radius r.
    // Return minimum r so every house is within r of at least one heater.

    /**
     * Sort both. For each house, binary search nearest heater. Take max distance.
     * Time O(n log n + m log m), Space O(1)
     */
    public static int findRadius(int[] houses, int[] heaters) {
        java.util.Arrays.sort(houses);
        java.util.Arrays.sort(heaters);
        int maxDist = 0;
        for (int h : houses) {
            int idx = java.util.Arrays.binarySearch(heaters, h);
            if (idx >= 0) continue;
            idx = -(idx + 1);
            int d = Integer.MAX_VALUE;
            if (idx < heaters.length) d = heaters[idx] - h;
            if (idx > 0) d = Math.min(d, h - heaters[idx - 1]);
            maxDist = Math.max(maxDist, d);
        }
        return maxDist;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Heaters");
        System.out.println("=".repeat(60));

        System.out.println("houses=[1,2,3], heaters=[2] -> " + findRadius(new int[]{1, 2, 3}, new int[]{2}));
        System.out.println("houses=[1,2,3,4], heaters=[1,4] -> " + findRadius(new int[]{1, 2, 3, 4}, new int[]{1, 4}));
    }
}

File: basic_dsa_two_hundred_fifty.py
"""
Basic DSA - Longest Increasing Subsequence (LIS)
Given an integer array nums, return the length of the longest strictly increasing subsequence.
"""


def length_of_lis(nums):
    """
    Patience sorting with binary search on tails.
    Time Complexity: O(n log n)
    Space Complexity: O(n)
    """
    tails = []
    for x in nums:
        lo, hi = 0, len(tails)
        while lo < hi:
            mid = (lo + hi) // 2
            if tails[mid] < x:
                lo = mid + 1
            else:
                hi = mid
        if lo == len(tails):
            tails.append(x)
        else:
            tails[lo] = x
    return len(tails)


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Longest Increasing Subsequence (LIS)")
    print("=" * 60)

    print(length_of_lis([10, 9, 2, 5, 3, 7, 101, 18]))  # 4
    print(length_of_lis([0, 1, 0, 3, 2, 3]))  # 4
